From 301420a8cc44ee457670e015adc3a08cfc4531a9 Mon Sep 17 00:00:00 2001
From: Jonathan Bell <jonathan@raspberrypi.com>
Date: Wed, 8 Jan 2025 14:58:37 +0000
Subject: [PATCH] DT: bcm2712: override supports-cqe to a cell

We want to be able to control the interop surface exposed by Command
Queueing across bcm2712 products to a more restrictive default, with
selectable disable and permissive behaviour.

Changing the bool to a cell lets it relay a tristate value.

Also add the override parameter to CM5 as CM5-lite may interface with
arbitrary eMMC or SD cards.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>
---
 arch/arm/boot/dts/overlays/README                 | 11 ++++++++---
 arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts  |  4 ++--
 arch/arm64/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi |  3 ++-
 arch/arm64/boot/dts/broadcom/bcm2712.dtsi         |  2 +-
 4 files changed, 13 insertions(+), 7 deletions(-)

--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -405,9 +405,14 @@ Params:
                                 non-lite SKU of CM4).
                                 (default "on")
 
-        sd_cqe                  Set to "off" to disable Command Queueing if you
-                                have an incompatible Class A2 SD card
-                                (Pi 5 only, default "on")
+        sd_cqe                  Modify Command Queuing behaviour on the main SD
+                                interface. Legal values are:
+                                0: disable CQ
+                                1: allow CQ for known-good SD A2 cards, and all
+                                   eMMC cards
+                                2: allow CQ for all SD A2 cards that aren't
+                                   known-bad, and all eMMC cards.
+                                (2712 only, default "1")
 
         sd_overclock            Clock (in MHz) to use when the MMC framework
                                 requests 50MHz
--- a/arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts
+++ b/arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts
@@ -365,7 +365,7 @@ dpi_16bit_gpio2:        &rp1_dpi_16bit_g
 	sd-uhs-sdr50;
 	sd-uhs-ddr50;
 	sd-uhs-sdr104;
-	supports-cqe;
+	supports-cqe = <1>;
 	cd-gpios = <&gio_aon 5 GPIO_ACTIVE_LOW>;
 	//no-1-8-v;
 	status = "okay";
@@ -745,6 +745,6 @@ spi10_cs_pins: &spi10_cs_gpio1 {};
 
 / {
 	__overrides__ {
-		sd_cqe = <&sdio1>, "supports-cqe?";
+		sd_cqe = <&sdio1>, "supports-cqe:0";
 	};
 };
--- a/arch/arm64/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi
+++ b/arch/arm64/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi
@@ -339,7 +339,7 @@ dpi_16bit_gpio2:        &rp1_dpi_16bit_g
 	mmc-hs400-1_8v;
 	mmc-hs400-enhanced-strobe;
 	broken-cd;
-	supports-cqe;
+	supports-cqe = <1>;
 	status = "okay";
 };
 
@@ -752,5 +752,6 @@ spi10_cs_pins: &spi10_cs_gpio1 {};
 			<&ant2>, "output-low?=on";
 		noanthogs = <&ant1>,"status=disabled",
 			<&ant2>, "status=disabled";
+		sd_cqe = <&sdio1>, "supports-cqe:0";
 	};
 };
--- a/arch/arm64/boot/dts/broadcom/bcm2712.dtsi
+++ b/arch/arm64/boot/dts/broadcom/bcm2712.dtsi
@@ -1210,7 +1210,7 @@
 			clocks = <&clk_emmc2>;
 			sdhci-caps-mask = <0x0000C000 0x0>;
 			sdhci-caps = <0x0 0x0>;
-			supports-cqe;
+			supports-cqe = <1>;
 			mmc-ddr-3_3v;
 			status = "disabled";
 		};
